### PR TITLE
Make variable substitution work as expected, PQ graphing upgrade, DataLink generation

### LINKED ISSUE #29: Template parameters (grafana variables) are only appended to the end of a query
When we add template parameters, I would like to be able to define the param wherever i want in the query but the params are appended to the end of the request. 

For example, I would like to be able to define a list of params in Grafana 

host = server1, server2, server3

and select server1 from the dropdown. 

with the following query:

dataset = 'accesslog' && $serverHost = '[[host]]'

I would expect. the body of the response to be 

data:"{"token":"token","queries":[{"startTime":1575658166918,"endTime":1575659917468,"buckets":1750,"filter":"dataset='accesslog' $serverHost == 'server1'","function":"mean(time)"}]}"

But instead, i get 

data:"{"token":"token","queries":[{"startTime":1575658166918,"endTime":1575659917468,"buckets":1750,"filter":"dataset='accesslog' $serverHost == '[[host]]' server1","function":"mean(time)"}]}"

### LINKED ISSUE #39: Graph more columns of a PowerQuery
Currently power queries are able to be graphed with the `Graph` visualization type, to do so you must set the x-axis mode to `Series`. This will result in the values of column 1 being used at the series names and acting as the x-axis labels, and the values of column 2 as the bar value. All other columns are not shown.

It would be good to show all the columns returned by a PQ.

### PR DESCRIPTION
**Pull request by yanscalyr**
A first pass run at changes needed for v2.

Changes to variable substitution will fix #29 and allow the same dropdown to filter by facet functionality as in Scalyr, assuming the query is set up correctly.
PQ graphing more or less already existed in a form that seems reasonable, but this changes it to support the graphing of more columns from the result set.

This resolves #39 
This resolves #29

### CODE REVIEW

**Comment by yanscalyr** [comment]
Note that this will break the earlier functionality of variables being added in to queries automatically as limits on facets of their name.

File: src/query_ctrl.js [hunk 1]
```diff
@@ -65,10 +76,53 @@ export class GenericDatasourceQueryCtrl extends QueryCtrl {
...
+    return str;
+  }
+
+  createDataLinkURL() {
+      if (this.target.queryType === this.queryTypes.STANDARD_QUERY) {
+        let dataLinkFilter = ""
+        if (this.target.queryText) {
+          const extractedVars = this.target.queryText.match(varRegex);
+          const varRegex = /\$(\w+)|\[\[([\s\S]+?)(?::(\w+))?\]\]|\${(\w+)(?:\.([^:^}]+))?(?::(\w+))?}/g
+          const queryWithoutVars = GenericDatasourceQueryCtrl.splitOnArrayElements(this.target.queryText, extractedVars);
```
[the remark this answers is not in the document]
Comment (yanscalyr) [reply src/query_ctrl.js]: The linter forced methods that could be static to be static, so I needed to call them this way. Other changes get rid of this any though.

File: src/query_ctrl.js [hunk 2]
```diff
@@ -65,10 +76,53 @@ export class GenericDatasourceQueryCtrl extends QueryCtrl {
...
 
+  static splitOnArrayElements(str, splitters) {
+    let result = [str];
+    if (splitters) {
+      for (let i = 0; i < splitters.length; i += 1) {
+        let subresult = [];
+        for (let j = 0; j < result.length; j += 1) {
+          subresult = subresult.concat(result[j].split(splitters[i]));
+        }
+        result = subresult;
```
[the remark this answers is not in the document]
Comment (yanscalyr) [reply src/query_ctrl.js]: It's setting `subresult` to itself concatenated with something else, think of it as a `+=` operation.

File: src/query_ctrl.js [hunk 3]
```diff
@@ -65,10 +76,53 @@ export class GenericDatasourceQueryCtrl extends QueryCtrl {
...
+    }
+    return str;
+  }
+
+  createDataLinkURL() {
+      if (this.target.queryType === this.queryTypes.STANDARD_QUERY) {
+        let dataLinkFilter = ""
+        if (this.target.queryText) {
+          const extractedVars = this.target.queryText.match(varRegex);
+          const varRegex = /\$(\w+)|\[\[([\s\S]+?)(?::(\w+))?\]\]|\${(\w+)(?:\.([^:^}]+))?(?::(\w+))?}/g
```
[the remark this answers is not in the document]
Comment (yanscalyr) [reply src/query_ctrl.js]: The regex was pulled directly from the Grafana github, and its purpose is to be the same as the one in Grafana. Any issues this regex has should be kept anyway so I don't think tests help. Though I should leave a comment linking back to where I got it.

File: src/query_ctrl.js (hunk 2 above)
[the remark this answers is not in the document]
Comment (guranjan) [reply src/query_ctrl.js]: right, but you're also resetting subresult and then assigning it to result in the outer loop. So the result value will be what ever was in the last iteration of the outer loop.

File: src/datasource.js [hunk 4]
```diff
@@ -132,14 +133,36 @@ export class GenericDatasource {
     });
   }
 
+  static interpolateVariable(value, variable) {
```
[the remark this answers is not in the document]
Comment (guranjan) [reply src/datasource.js]: 🌮

File: src/query_ctrl.js (hunk 2 above)
[the remark this answers is not in the document]
Comment (yanscalyr) [reply src/query_ctrl.js]: But the value of `subresult` depends on the value of result from the previous loop, we split every element of `result` and put all those into `subresult`. On the next outer loop the value of result will be the previous result but split on the (i-1)th splitter.

[the remark this answers is not in the document]
Comment (guranjan) [reply src/query_ctrl.js]: Ok. I walked through the code manually. Makes sense. Here's a much simpler implementation
`return str.split(new RegExp(splitter.join("|"))`

[the remark this answers is not in the document]
Comment (guranjan) [reply src/query_ctrl.js]: I'll leave it upto you if you want to make the change or not 👍

[the remark this answers is not in the document]
Comment (yanscalyr) [reply src/query_ctrl.js]: That makes sense, my mind never goes to regexs unless it needs too. I'll make the change.

[the remark this answers is not in the document]
Comment (yanscalyr) [reply src/query_ctrl.js]: Actually, the values in `splitters` can include characters like `$` or anything really, and we want to split on the string literal and not have that stuff be interpreted as a regex. Is there some way to easily put them in that regex as a literal?

[the remark this answers is not in the document]
Comment (guranjan) [reply src/query_ctrl.js]: You can escape it with `\`

[the remark this answers is not in the document]
Comment (yanscalyr) [reply src/query_ctrl.js]: I think I'll leave it as it is, messing around with trying to escape everything right will only lead to tears.

**Comment by yanscalyr** [comment]
> Also the build should be copying README.md to dist/README.md I didn't see that happen. Could you take a look at that.

I didn't have the `dist` directory added to be checked in, my bad. Should be in now.

**Pull request merged**